src/sentry/integrations/slack/endpoints/action.py

```diff
@@ -338,3 +353,72 @@ def post(self, request: Request) -> Response:
         body = self.construct_reply(attachment, is_message=self.is_message(data))
 
         return self.respond(body)
+
+    def handle_member_approval(self, slack_request: SlackActionRequest):
+        try:
+            identity = slack_request.get_identity()
+        except IdentityProvider.DoesNotExist:
+            return self.respond(status=403)
+
+        member_id = slack_request.callback_data["member_id"]
+        try:
+            # endpoint will check to make sure member is still in a state to be invited
+            member = OrganizationMember.objects.get(id=member_id)
+        except OrganizationMember.DoesNotExist:
+            return self.respond(status=404)
+
+        org_slug = member.organization.slug
+        url = reverse(
+            "sentry-api-0-organization-invite-request-detail",
+            args=[org_slug, member_id],
+        )
+
+        approve_member = slack_request.action_option == "approve_member"
+
+        # payload and endpoint differ if we reject or accept
+        action_data = {}
+        method = "DELETE"
+        if approve_member:
+            method = "PUT"
+            action_data = {
+                "approve": 1,
+            }
+
+        member_email = member.email
+        original_status = member.invite_status
+
+        try:
+            client.request(
```

Comment:
We should not be making API calls inside other API calls. You should move the member approval and invite deletion code into helpers and call those helpers here.

Reply:
@mgaeta How can I avoid calling this API though? Or do you mean just moving `client.request` into a helper function?

Reply:
No you should take the functionality that you want to use and extract it to a helper like
```py
def approve_member(
    organization: Organization, 
    member: OrganizationMemer,
    request: Request,
) -> None:
    member.approve_invite()
    member.save()
    
    if settings.SENTRY_ENABLE_INVITES:
        member.send_invite_email()
        member_invited.send_robust(
            member=member,
            user=request.user,
            sender=self,
            referrer=request.data.get("referrer"),
        )
    
    self.create_audit_entry(
        request=request,
        organization_id=organization.id,
        target_object=member.id,
        data=member.get_audit_log_data(),
        event=AuditLogEntryEvent.MEMBER_INVITE
        if settings.SENTRY_ENABLE_INVITES
        else AuditLogEntryEvent.MEMBER_ADD,
    )
```

Reply:
@mgaeta I see what you're saying. But calling a Sentry API from a webhook is something we do quite often. See this example in [release hooks](https://github.com/getsentry/sentry/blob/master/src/sentry/web/frontend/release_webhook.py#L53) which written by David Cramer. Note, we don't actually call the APIs when we do this with the right client: https://github.com/getsentry/sentry/blob/master/src/sentry/api/client.py#L99. But it's convenient to mock calling the APIs because we have API validation. If I didn't use the APIs here, I would either have to duplicate logic such as [this](https://github.com/getsentry/sentry/blob/master/src/sentry/api/endpoints/organization_invite_request_details.py#L22-L44) in the endpoint. So using APIs like this isn't just OK, it's actually preferred.

Reply:
We also directly use the internal API client here https://github.com/getsentry/sentry/blob/53f99d3d2a8b194823eed24ce79d570ccc765a23/src/sentry/integrations/slack/unfurl/discover.py#L195-L200

Reply:
We should stop using this pattern and remove it where we can. If you need validation, you should extract that logic into a helper as well. The method `validate_approve()` is also not a good example of how to do a permissions check in an API.

Reply:
@mgaeta There is more validation than just the serializer such as validating that the requester is part of the org of the member request. I'd have to duplicate that logic as well which would be difficult to maintain if we ever changed the underlying logic. Could you explain _why_ this pattern should be removed?

Reply:
@mgaeta Ok...talked to Dan and he agrees with you, I'll switch to not using APIs